moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -0,0 +1,1020 @@
+}
+
+bool Trajectory::getNextVelocitySwitchingPoint(double path_pos, TrajectoryStep& next_switching_point,
+                                               double& before_acceleration, double& after_acceleration)
+{
+  const double step_size = 0.001;
+  const double accuracy = 0.000001;
+
+  bool start = false;
+  path_pos -= step_size;
+  do
+  {
+    path_pos += step_size;
+
+    if (getMinMaxPhaseSlope(path_pos, getVelocityMaxPathVelocity(path_pos), false) >=
+        getVelocityMaxPathVelocityDeriv(path_pos))
+    {
+      start = true;
+    }
+  } while ((!start ||
+            getMinMaxPhaseSlope(path_pos, getVelocityMaxPathVelocity(path_pos), false) >
+                getVelocityMaxPathVelocityDeriv(path_pos)) &&
+           path_pos < path_.getLength());
+
+  if (path_pos >= path_.getLength())
+  {
+    return true;  // end of trajectory reached
+  }
+
+  double before_path_pos = path_pos - step_size;
+  double after_path_pos = path_pos;
+  while (after_path_pos - before_path_pos > accuracy)
+  {
+    path_pos = (before_path_pos + after_path_pos) / 2.0;
+    if (getMinMaxPhaseSlope(path_pos, getVelocityMaxPathVelocity(path_pos), false) >
+        getVelocityMaxPathVelocityDeriv(path_pos))
+    {
+      before_path_pos = path_pos;
+    }
+    else
+    {
+      after_path_pos = path_pos;
+    }
+  }
+
+  before_acceleration = getMinMaxPathAcceleration(before_path_pos, getVelocityMaxPathVelocity(before_path_pos), false);
+  after_acceleration = getMinMaxPathAcceleration(after_path_pos, getVelocityMaxPathVelocity(after_path_pos), true);
+  next_switching_point = TrajectoryStep(after_path_pos, getVelocityMaxPathVelocity(after_path_pos));
+  return false;
+}
+
+// Returns true if end of path is reached
+bool Trajectory::integrateForward(std::list<TrajectoryStep>& trajectory, double acceleration)
+{
+  double path_pos = trajectory.back().path_pos_;
+  double path_vel = trajectory.back().path_vel_;
+
+  std::list<std::pair<double, bool>> switching_points = path_.getSwitchingPoints();
+  std::list<std::pair<double, bool>>::iterator next_discontinuity = switching_points.begin();
+
+  while (true)
+  {
+    while ((next_discontinuity != switching_points.end()) &&
+           (next_discontinuity->first <= path_pos || !next_discontinuity->second))
+    {
+      ++next_discontinuity;
+    }
+
+    double old_path_pos = path_pos;
+    double old_path_vel = path_vel;
+
+    path_vel += time_step_ * acceleration;
+    path_pos += time_step_ * 0.5 * (old_path_vel + path_vel);
+
+    if (next_discontinuity != switching_points.end() && path_pos > next_discontinuity->first)
+    {
+      path_vel = old_path_vel +
+                 (next_discontinuity->first - old_path_pos) * (path_vel - old_path_vel) / (path_pos - old_path_pos);
+      path_pos = next_discontinuity->first;
+    }
+
+    if (path_pos > path_.getLength())
+    {
+      trajectory.push_back(TrajectoryStep(path_pos, path_vel));
+      return true;
+    }
+    else if (path_vel < 0.0)
```

Comment:
Hi, I know this is as already merged code but is this right? My project has heavy deacceleration and my path points generated from compute_cartesian (which works with iterative spline) does not work with totp. I get the error msg : "Error while integrating forward: Negative path velocity". Why cant the path velocity be negative (sorry haven't looked through the code enough). Please let me know if this seems like a valid query, I'll open a git issue then.